File: mcp-server-scaffold/templates/python/template-tool.py
"""
{{TOOL_NAME}} Tool Handler

{{DESCRIPTION}}
"""

from pydantic import BaseModel, Field
from mcp.types import TextContent, ErrorCode, McpError
from typing import Any


# === INPUT SCHEMA ===

class {{TOOL_NAME_PASCAL}}Input(BaseModel):
    """Input parameters for {{TOOL_NAME}} tool."""
    
    # Define your input parameters here
    # Example:
    # location: str = Field(..., description="City name or coordinates")
    # units: str = Field(default="celsius", description="Temperature units (celsius/fahrenheit)")
    
    {{PARAMS}}


# === TOOL DEFINITION ===

{{TOOL_NAME_UPPER}}_TOOL = {
    "name": "{{TOOL_NAME}}",
    "description": "{{DESCRIPTION}}",
    "inputSchema": {
        "type": "object",
        "properties": {
            # Mirror Pydantic schema in JSON Schema format
            # Example:
            # "location": {
            #     "type": "string",
            #     "description": "City name or coordinates",
            # },
            # "units": {
            #     "type": "string",
            #     "enum": ["celsius", "fahrenheit"],
            #     "default": "celsius",
            #     "description": "Temperature units",
            # },
            
            {{PROPERTIES}}
        },
        "required": [{{REQUIRED_PARAMS}}],
    },
}


# === HANDLER ===

async def handle_{{TOOL_NAME_SNAKE}}(
    arguments: dict[str, Any]
) -> list[TextContent]:
    """
    Handle {{TOOL_NAME}} tool execution.
    
    Args:
        arguments: Tool input parameters
        
    Returns:
        List of content items to return to the client
    """
    # Validate input
    try:
        params = {{TOOL_NAME_PASCAL}}Input(**arguments)
    except Exception as e:
        raise McpError(
            ErrorCode.InvalidParams,
            f"Invalid parameters: {str(e)}"
        )
    
    try:
        # === IMPLEMENT YOUR TOOL LOGIC HERE ===
        
        # Example:
        # result = await fetch_weather(params.location, params.units)
        
        result = await execute_{{TOOL_NAME_SNAKE}}(params)
        
        return [
            TextContent(
                type="text",
                text=format_result(result),
            )
        ]
        
    except McpError:
        raise
    except Exception as e:
        import logging
        logging.error(f"{{TOOL_NAME}} tool error: {e}")
        
        raise McpError(
            ErrorCode.InternalError,
            f"Failed to execute {{TOOL_NAME}}: {str(e)}"
        )


# === IMPLEMENTATION ===

async def execute_{{TOOL_NAME_SNAKE}}(
    params: {{TOOL_NAME_PASCAL}}Input
) -> Any:
    """
    Execute the actual tool logic.
    
    TODO: Implement the actual functionality here.
    """
    raise NotImplementedError("execute_{{TOOL_NAME_SNAKE}} not implemented")


def format_result(result: Any) -> str:
    """
    Format the result for display.
    
    TODO: Customize formatting based on your result type.
    """
    import json
    
    if isinstance(result, str):
        return result
    
    return json.dumps(result, indent=2, default=str)

File: skills/mcp-server-scaffold/templates/typescript/template-tool.ts
import { z } from "zod";
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";

/**
 * {{TOOL_NAME}} Tool Handler
 * 
 * {{DESCRIPTION}}
 */

// === INPUT SCHEMA ===

export const {{TOOL_NAME_CAMEL}}Schema = z.object({
  // Define your input parameters here
  // Example:
  // location: z.string().describe("City name or coordinates"),
  // units: z.enum(["celsius", "fahrenheit"]).default("celsius"),
  
  {{PARAMS}}
});

export type {{TOOL_NAME_PASCAL}}Input = z.infer<typeof {{TOOL_NAME_CAMEL}}Schema>;

// === TOOL DEFINITION ===

export const {{TOOL_NAME}}Definition = {
  name: "{{TOOL_NAME}}",
  description: "{{DESCRIPTION}}",
  inputSchema: {
    type: "object" as const,
    properties: {
      // Mirror Zod schema in JSON Schema format
      // Example:
      // location: {
      //   type: "string",
      //   description: "City name or coordinates",
      // },
      // units: {
      //   type: "string",
      //   enum: ["celsius", "fahrenheit"],
      //   description: "Temperature units",
      // },
      
      {{PROPERTIES}}
    },
    required: [{{REQUIRED_PARAMS}}],
  },
};

// === HANDLER ===

export async function handle{{TOOL_NAME_PASCAL}}(
  args: unknown
): Promise<{ content: Array<{ type: "text"; text: string }> }> {
  // Validate input
  const parseResult = {{TOOL_NAME_CAMEL}}Schema.safeParse(args);
  if (!parseResult.success) {
    throw new McpError(
      ErrorCode.InvalidParams,
      `Invalid parameters: ${parseResult.error.message}`
    );
  }

  const params = parseResult.data;

  try {
    // === IMPLEMENT YOUR TOOL LOGIC HERE ===
    
    // Example:
    // const result = await fetchWeather(params.location, params.units);
    
    const result = await execute{{TOOL_NAME_PASCAL}}(params);
    
    return {
      content: [
        {
          type: "text" as const,
          text: formatResult(result),
        },
      ],
    };
  } catch (error) {
    console.error(`{{TOOL_NAME}} tool error:`, error);
    
    throw new McpError(
      ErrorCode.InternalError,
      `Failed to execute {{TOOL_NAME}}: ${error}`
    );
  }
}

// === IMPLEMENTATION ===

async function execute{{TOOL_NAME_PASCAL}}(
  params: {{TOOL_NAME_PASCAL}}Input
): Promise<unknown> {
  // TODO: Implement the actual tool logic
  
  throw new Error("Not implemented: execute{{TOOL_NAME_PASCAL}}");
}

function formatResult(result: unknown): string {
  // TODO: Format the result for display
  
  return JSON.stringify(result, null, 2);
}
